python/ciqueue/distributed.py

```diff
@@ -0,0 +1,165 @@
+import os
+import time
+import math
+from redis import ConnectionError
+
+from ciqueue.static import Static
+
+
+class LostMaster(Exception):
+    pass
+
+
+class Base(object):
+    def __init__(self, redis, build_id):
+        self.redis = redis
+        self.build_id = str(build_id)
+        self.is_master = False
+        self._scripts = {}
+
+    def key(self, *args):
+        return ':'.join(['build', self.build_id] + map(str, args))
+
+    def wait_for_master(self, timeout=10):
+        if self.is_master:
+            return True
+
+        for i in range(timeout * 10 + 1):
+            master_status = self._master_status()
+            if master_status in ['ready', 'finished']:
+                return True
+            time.sleep(0.1)
+
+        raise LostMaster(
+            "The master worker is still `" +
+            repr(master_status) +
+            "` after 10 seconds waiting.")
+
+    def _master_status(self):
+        return self.redis.get(self.key('master-status'))
+
+    def __len__(self):
+        transaction = self.redis.pipeline(transaction=True)
+        transaction.llen(self.key('queue'))
+        transaction.zcard(self.key('running'))
+        return sum(transaction.execute())
+
+    @property
+    def progress(self):
+        return self.total - len(self)
+
+
+class Worker(Base):
+    def __init__(self, tests, worker_id, redis, build_id,
+                 timeout, max_requeues=0, requeue_tolerance=0):
+        super(Worker, self).__init__(redis=redis, build_id=build_id)
+        self.timeout = timeout
+        self.total = len(tests)
+        self.max_requeues = max_requeues
+        self.global_max_requeues = math.ceil(len(tests) * requeue_tolerance)
+        self.worker_id = worker_id
+        self.shutdown_required = False
+        self._push(tests)
+
+    def __iter__(self):
+        self.wait_for_master()
+
+        while not self.shutdown_required and len(self):
+            test = self._reserve()
+            if test:
+                yield test
+            else:
+                return
+                time.sleep(0.05)
+
+    def shutdown(self):
+        self.shutdown_required = True
+
+    def acknowledge(self, test):
+        return self._eval_script(
+            'acknowledge',
+            keys=[self.key('running'), self.key('processed')],
+            args=[test],
+        ) == 1
+
+    def requeue(self, test, offset=42):
+        return self._eval_script(
+            'requeue',
+            keys=[
+                self.key('processed'),
+                self.key('requeues-count'),
+                self.key('queue'),
+                self.key('running')],
+            args=[self.max_requeues, self.global_max_requeues, test, offset],
+        ) == 1
```

Comment:
We could raise indeed, but booleans work quite well here:
  - True -> the test was requeued
  - False -> the test wasn't requeued (because the limit was reached already)